File: 0564/Solution.js
/**
 * @param {string} n
 * @return {string}
 */
var nearestPalindromic = function(n) {
    let num = BigInt(n);
    if(num <= 10n) return (num - 1n).toString();
    if(num === 11n) return "9";
    let length = n.length;

    let leftHalf = BigInt(n.slice(0, (length + 1) / 2));
    
    let candidates = [
        genPalFromLeftHalf(leftHalf, length % 2 == 0),
        genPalFromLeftHalf(leftHalf - 1n, length % 2 == 0),
        genPalFromLeftHalf(leftHalf + 1n, length % 2 == 0),
        BigInt(10n ** BigInt(length - 1)) - 1n,
        BigInt(10n ** BigInt(length)) + 1n,
    ]

  

    let minDif =BigInt(Number.MAX_SAFE_INTEGER);
    let res = 0n;

    for(let candidate of candidates){
        if(candidate === num) continue;
        let dif = candidate > num ? candidate - num : num - candidate;
        if(dif < minDif || (dif === minDif && candidate < res)){
            res = candidate;
            minDif = dif;
        }
    }
    return res.toString();
};

function genPalFromLeftHalf(leftHalf, even){
    let res = leftHalf;
    if(!even) leftHalf = leftHalf / 10n;
    while(leftHalf > 0n){
        res = res * 10n + leftHalf % 10n;
        leftHalf /= 10n;
    }
    return res;
}

File: 0564/Solution.java
class Solution {
    public String nearestPalindromic(String n) {
       long num = Long.parseLong(n);
       if(num <= 10) return String.valueOf(num - 1);
       if(num == 11) return "9";

       int length = n.length();
       long leftHalf = Long.parseLong(n.substring(0, (length + 1) / 2));

       long[] candidates = new long[5];

       candidates[0] = generatePalFromLeft(leftHalf -1, length % 2 == 0);
       candidates[1] = generatePalFromLeft(leftHalf, length % 2 == 0);
       candidates[2] = generatePalFromLeft(leftHalf +1, length % 2 == 0);
       candidates[3] = (long) Math.pow(10, length -1) - 1;
       candidates[4] = (long) Math.pow(10, length)+ 1;

       long nearestPal = 0;
       long minDif = Long.MAX_VALUE;

       for(long candidate : candidates){
            if(candidate == num) continue;
            long dif = Math.abs(candidate - num);
            if(dif < minDif || (dif == minDif && candidate < nearestPal)){
                minDif = dif;
                nearestPal = candidate;
            }
       }
       return String.valueOf(nearestPal);

    }
    private long generatePalFromLeft(long leftHalf, boolean even){
        long pal = leftHalf;
        if(!even) leftHalf /= 10;
        while(leftHalf > 0){
            pal = pal * 10 + leftHalf % 10;
            leftHalf /= 10;
        }
        return pal;
    }
}

// time O(1)
// space O(1)
